### PR TITLE
🐛 avoid in-place capbk mutation, duplicate files

### PR DESCRIPTION
**Pull request by alexeldeib**
**What this PR does / why we need it**:

we write the certificate files twice into cloud init, once as writeFiles and again as additionalFiles which get written back to writeFiles.

we also mutate the kubeadm config spec in place when converting contentFrom -> content. avoid doing that.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

/hold

for some tests

### CODE REVIEW

File: bootstrap/kubeadm/controllers/kubeadmconfig_controller.go [hunk 1]
```diff
@@ -578,20 +578,35 @@ func (r *KubeadmConfigReconciler) joinControlplane(ctx context.Context, scope *S
...
 			}
 			in.ContentFrom = nil
 			in.Content = string(data)
-			collected[i] = in
+			collected[index] = in
+			index++
 		}
+	}
 
+	for i := range merge {
```
Comment (CecileRobertMichon) [inline bootstrap/kubeadm/controllers/kubeadmconfig_controller.go:598]: is this unit tested?

Comment (CecileRobertMichon) [reply bootstrap/kubeadm/controllers/kubeadmconfig_controller.go:598]: edit: just saw you have a hold for tests, nvm

**Comment by alexeldeib** [comment]
I think for the join case, we take https://github.com/kubernetes-sigs/cluster-api/blob/2afc70d32567682f93468033433e72fd428fc8f9/bootstrap/kubeadm/controllers/kubeadmconfig_controller.go#L531
and plumb it through to https://github.com/kubernetes-sigs/cluster-api/blob/2afc70d32567682f93468033433e72fd428fc8f9/bootstrap/kubeadm/internal/cloudinit/controlplane_join.go#L57-L59 which handles the same logic for us: https://github.com/kubernetes-sigs/cluster-api/blob/2afc70d32567682f93468033433e72fd428fc8f9/bootstrap/kubeadm/internal/cloudinit/cloudinit.go#L57

will see if I can align these, or if there's a reason not to 👍

**Comment by alexeldeib** [comment]
I guess `prepare()` only really makes sense for join, since this code isn't helpful for init: https://github.com/kubernetes-sigs/cluster-api/blob/2afc70d32567682f93468033433e72fd428fc8f9/bootstrap/kubeadm/internal/cloudinit/cloudinit.go#L58-L66

**Comment by fabriziopandini** [comment]
@alexeldeib thanks for investigating this.
I think we are ok with prepare for join (eventually fixing this is out of scope of this PR)

File: bootstrap/kubeadm/controllers/kubeadmconfig_controller.go [hunk 2]
```diff
@@ -578,20 +578,35 @@ func (r *KubeadmConfigReconciler) joinControlplane(ctx context.Context, scope *S
 // resolveFiles maps .Spec.Files into cloudinit.Files, resolving any object references
 // along the way.
 func (r *KubeadmConfigReconciler) resolveFiles(ctx context.Context, cfg *bootstrapv1.KubeadmConfig, merge ...bootstrapv1.File) ([]bootstrapv1.File, error) {
-	collected := append(cfg.Spec.Files, merge...)
+	collected := make([]bootstrapv1.File, len(cfg.Spec.Files)+len(merge))
```
Comment (vincepri) [inline bootstrap/kubeadm/controllers/kubeadmconfig_controller.go:581]:
```suggestion
	collected := make([]bootstrapv1.File, 0, len(cfg.Spec.Files)+len(merge))
```

Then we don't need `index`, but we can just `append`

**Comment by alexeldeib** [comment]
@CecileRobertMichon added tests

File: bootstrap/kubeadm/controllers/kubeadmconfig_controller.go (hunk 2 above)
Comment (alexeldeib) [reply bootstrap/kubeadm/controllers/kubeadmconfig_controller.go]: not actually merging anything anymore, so got rid of this altogether and switched to append

File: bootstrap/kubeadm/controllers/kubeadmconfig_controller_test.go [hunk 3]
```diff
@@ -1495,6 +1495,148 @@ func TestKubeadmConfigReconciler_Reconcile_PatchWhenErrorOccurred(t *testing.T)
...
+				{
+					Content:     "foo",
+					Path:        "/path",
+					Owner:       "root:root",
+					Permissions: "0600",
+				},
+			},
+			objects: []runtime.Object{testSecret},
+		},
+		"multiple files should work correctly": {
```
Comment (CecileRobertMichon) [inline bootstrap/kubeadm/controllers/kubeadmconfig_controller_test.go:1563]: was about to hit enter on a comment asking exactly for this :D

File: bootstrap/kubeadm/controllers/kubeadmconfig_controller_test.go [hunk 4]
```diff
@@ -1495,6 +1495,148 @@ func TestKubeadmConfigReconciler_Reconcile_PatchWhenErrorOccurred(t *testing.T)
...
+			g := NewWithT(t)
+
+			myclient := helpers.NewFakeClientWithScheme(setupScheme(), tc.objects...)
+			k := &KubeadmConfigReconciler{
+				Log:             log.Log,
+				Client:          myclient,
+				KubeadmInitLock: &myInitLocker{},
+			}
+
+			// make a list of files we expect to be source from secrets
```
Comment (alexeldeib) [inline bootstrap/kubeadm/controllers/kubeadmconfig_controller_test.go:1616]: spelling typo, source -> sourced

**Pull request merged**